### ISSUE TITLE
Add round to significant figures in package 

### ISSUE THREAD

**Issue by bc118** [numpy/numpy#25848]
### Proposed new feature or change:

Code Suggestion:

def round_to_sig_figs(value, sig_figs=3):
   """Round a number to the selected number of significant figures.

   Round a number to the selected number of significant figures.

   Parameters
   ----------
   value: int or float
       The number that will be rounded to the selected number of
       significant figures.
   sig_figs: int, default=3
       The number significant figures that the 'value' variable
       will be rounded too. If sig_fig=0, it will return 0.0.

   Returns
   -------
   value_rounded_to_sig_figs: float
       The input 'value' variable rounded to the selected number
       significant figures. If sig_fig=0, it will return 0.0.
   """
   if value == 0:
      value_rounded_to_sig_figs = 0

   else:
      value_rounded_to_sig_figs = float(
         round(value, sig_figs - int(math.floor(math.log10(abs(value)))) - 1)
      )

   return value_rounded_to_sig_figs

**Comment by rgommers** [comment]
Thanks for the suggestion @bc118. This is a utility that can be written in a few lines of Python code on top of `np.round`, so it looks to me like it doesn't meet the bar for inclusion in numpy.

Additionally, would this function even be useful for arrays, or only for floats?

**Comment by bc118** [comment]
IMO it should be included, as it is helpful for statistics and all applications, and it is annoying to add in every code that people write. Many people in the community wonder why this is not added, along with other things.  

If you want to round arrays or floats, feel free to add this.  I am not opposed to anything, I am just suggesting it be added in general.

**Comment by bc118** [comment]
Why have any other round feature, like decimals?

**Comment by bc118** [comment]
This will be my last time contributing to this repo

**Comment by mattip** [comment]
We follow the python lead, which has `__round__` in its [object model](https://docs.python.org/3/reference/datamodel.html#object.__round__). The usual way to approach adding new functionality in NumPy, before the Array API, was to [ask the mailing list](https://numpy.org/devdocs/dev/development_workflow.html#get-the-mailing-list-s-opinion) for comment. These days we tend to defer to the Array API, whose [round](https://data-apis.org/array-api/latest/API_specification/generated/array_api.round.html) does not even have a `decimals` kwarg, much less a `significant_digits` one.

**Comment by OyiboRivers** [comment]
@bc118 "significant figures" is a useful function, particularly where formatting matters.
Integrating a feature like this into numpy could indeed reduce repetitive implementations.

However, it's crucial to ensure that such a function covers all necessary cases.
Code examples (i.e. on stackoverflow) are sometimes error prone and in this specific case you could add a check for np.isfinite(value) to catch nan & inf values. You could add a remark that rounding rule is ties to equal (i.e. 2.5->2, 3.5->4). 
There are probably even more things to consider and there is a whole Python package called "rounders" dedicated to this topic.
https://github.com/mdickinson/rounders/tree/main

If a function like "significant figures" is used in many packages then robustness and reliability is paramount.
This would make the case to include such functions into the Python math standard library or Numpy's library.

The discussion around a similar feature request for Python's math library shows the complexity involved.
https://discuss.python.org/t/rounding-to-significant-figures-feature-request-for-math-library/16395/87

The fact that "significant figures" didn't make it over the math library bar suggests that the criteria for inclusion are beyond just utility reasons.
Nevertheless, feature requests like the "significant figures" function are valuable contributions.

@mattip given that the Array API guides the behavior of array methods, how does it impact the development and inclusion of standalone functions in numpy?
Are there specific considerations or guidelines that numpy follows to ensure consistency between array methods and standalone functions? What are numpy's main standards for inclusion of standalone functions?

I appreciate any insights.

Best regards, Oyibo

**Comment by rgommers** [comment]
> given that the Array API guides the behavior of array methods, how does it impact the development and inclusion of standalone functions in numpy?

I don't think it does in cases like this feature request. This is certainly too niche for a standard. Whether it deserves inclusion in `numpy` or `numpy.lib.xxx` can still be decided by the numpy team just fine. If there's any change, it's that the bar (which was already high) went up a bit more with all the care we've taken with the Python API cleanups and array API standard addition in NumPy 2.0

> If a function like "significant figures" is used in many packages then robustness and reliability is paramount.
This would make the case to include such functions into the Python math standard library or Numpy's library.

One could also say that if there are _that_ many ways to do it, pointing users at that `rounders` package is just fine. 

> Are there specific considerations or guidelines that numpy follows to ensure consistency between array methods and standalone functions? What are numpy's main standards for inclusion of standalone functions?

ndarray methods: we don't want any more, adding a new method will be highly exceptional. Functions are preferred.

For functions, they should at least be:
1. widely applicable
2. usually nontrivial to implement (if it's a few lines of pure Python/NumPy code, we tend to avoid adding it)
3. fit well in NumPy's existing scope

**Comment by rkern** [comment]
Personally, I'm less interested in adding a function that rounds the numbers themselves (that is, a function that takes an `ArrayLike` and returns an `ArrayLike`).

However, I'd love to have the [option to _format_](https://numpy.org/doc/stable/reference/generated/numpy.printoptions.html) float scalars and arrays rounded to a given number of significant digits rather than number digits after the decimal place. And preferably with [engineering notation](https://en.wikipedia.org/wiki/Engineering_notation).

**Comment by mattip** [comment]
Closing as "not planned"
